### DISCUSSION TITLE
psycopg2/_psycopg.cpython-312-x86_64-linux-gnu.so: undefined symbol: pg_strerror

### DISCUSSION THREAD

**Discussion by minhio** [psycopg/psycopg2#1713] [Q&A]
i'm trying to build psycopg2 for lambda runtime, got the following error that i can't seems to figure out, please help!

i'm packaging the psycopg2 built artifact including the libpq.* files from the postgresql build into the lambda runtime, and got the following error

`[ERROR] Runtime.ImportModuleError: Unable to import module 'main': /opt/python/psycopg2/_psycopg.cpython-312-x86_64-linux-gnu.so: undefined symbol: pg_strerror`

here's how i'm building psycopg2 https://github.com/minhio/psycopg2-lambda/blob/b4d8334aab37547b667080394c782d740cdd2a0f/Dockerfile

what am i missing? am i not building it correctly?

**Comment by dvarrazzo** [comment]
If someone else answers you, that's great. I personally don't have experience with lambda, and, because it is a paid-for proprietary platform, I cannot learn it for free, but only for a paid-for consultancy.

**Reply by minhio** [reply]
> If someone else answers you, that's great. I personally don't have experience with lambda, and, because it is a paid-for proprietary platform, I cannot learn it for free, but only for a paid-for consultancy.

i can ask this in a non-lambda specific way as this question is mainly around building psycopg2 from source and runtime dependencies

build this to reproduce the error

```Dockerfile
FROM fedora:39

ENV POSTGRES_VER=16.3
ENV PSYCOPG_VER=2.9.9

RUN yum update -y && yum upgrade -y \
    && yum install -y \
        tar \
        gcc \
        python3-devel \
        openssl-devel \
    && yum clean all \
    && rm -rf /var/cache/yum

# download postgres
WORKDIR /tmp
RUN curl -fsSL -o postgresql-${POSTGRES_VER}.tar.gz https://ftp.postgresql.org/pub/source/v${POSTGRES_VER}/postgresql-${POSTGRES_VER}.tar.gz \
    && tar -zxf postgresql-${POSTGRES_VER}.tar.gz

# build postgres
WORKDIR /tmp/postgresql-${POSTGRES_VER}
RUN ./configure --without-readline --without-icu --without-zlib --with-ssl=openssl \
    && make world-bin \
    && make install-world-bin

# download psycopg2
WORKDIR /tmp
RUN curl -fsSL -o psycopg2-${PSYCOPG_VER}.tar.gz https://github.com/psycopg/psycopg2/archive/refs/tags/${PSYCOPG_VER}.tar.gz \
    && tar -zxf psycopg2-${PSYCOPG_VER}.tar.gz

# build psycopg2
WORKDIR /tmp/psycopg2-${PSYCOPG_VER}
RUN python3 setup.py build_ext \
    --pg-config=/usr/local/pgsql/bin/pg_config \
    --static-libpq \
    --libraries=ssl,crypto \
    build

ENV LD_LIBRARY_PATH=/usr/local/pgsql/lib
ENV PATH=/usr/local/pgsql/bin:$PATH

RUN python3 setup.py install

RUN python3 -c "import psycopg2"

WORKDIR /

ENTRYPOINT [ "" ]
```

**Comment by minhio** [comment]
@dvarrazzo, removing `--static-libpq` or using `pip install psycopg2` seems to work, is static linking libpq not working?
